Starting specs 1 and 2 (#1)
pyviz/spec: spec_03.md

- Spec: 3
- Title: Methods on children of figures
- Version: 0.1
- Last-Modified: 2019-07-19
- Authors: [Jon Mease](jon.mease@gmail.com),
           [Julia Signell](jsignell@gmail.com),
- Status: Active
- Type: Standards
- Content-Type: text/markdown
- Created: 2019-07-19

## Abstract
This spec describes a proposed shared API for accessing objects within the figure hierarchy from various plotting libraries.

> This spec uses **figure** to refer to any object that has a visible representation but which can also be saved, exported, etc. This includes generic visualizations or animation that may or may not be a "plot" (axes, ticks, etc)?

Reviewer: ```suggestion
This spec uses **figure** to refer to any object that has a visible representation but which can also be saved, exported, etc. This includes generic visualizations or animations that may or may not be considered a "plot" with axes, ticks, and so on.
```

When referring to parts of a figure (such as axes and ticks) we'll use the term: **child**.

When detailing a method, this spec uses FIGURE to refer to any specific figure object and LIBRARY to refer to any specific visualization library, and CHILD to refer to the child of a figure.

## If a method doesn't make sense

> If a given operation doesn't make sense for that library, then it can satisfy the spec by simply having that method return a message to that effect ("JSON support not available in LIBRARY").

Reviewer: ```suggestion
If a given operation doesn't make sense for that library, then it can satisfy the spec by simply having that method return a message to that effect ("Unsupported: LIBRARY does not provide JSON output").
```

## Accessing root figure
Every child should know what figure it belongs to.

### Proposal

> Every object in the figure hierarchy should have a `.root` property that returns the parent figure, or if deeply nested, the grand or great-grandparent figure. If the object doesn't belong to a figure then it should return `None`.

Reviewer: ```suggestion
Every object in the figure hierarchy should have a `.root` property that returns the top-level parent figure. A deeply nested child object would not return its immediate parent, but the highest-level parent it is aware of (which may be its parent, grandparent, great-grandparent, etc., depending on how deeply it is nested). If the object doesn't belong to a figure then it should return `None`.
```

Here the intention is to return the ultimate, not the proximal, owner?  Under that assumption, I've tried to make that clear.

Should it really return None if it doesn't belong to a figure, or itself?  I'd vote for returning itself on principle, but I'm not sure how this support is meant to be used in practice.

Author: I think it is helpful to imagine a colorbar or some other part of a plot that isn't itself a figure, but would almost always belong to a figure. I think if that object were an orphan, then it makes sense to return that its root is `None`.
